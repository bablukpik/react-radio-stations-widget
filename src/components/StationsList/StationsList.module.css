.stationListWrapper {
  width: 100%;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  padding-left: 35px;
  padding-right: 35px;
  padding-bottom: 63px;
  padding-top: 54px;
  height: 544px;
  overflow: auto;
  scroll-behavior: smooth;
  background-color: #2f2f3a;
}

.stationListWrapper::-webkit-scrollbar {
  width: 0 !important;
  display: none;
}

.stationList {
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  list-style-type: none;
}
