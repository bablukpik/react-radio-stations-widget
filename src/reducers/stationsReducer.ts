export interface StationItem {
  name: string;
  id: string;
  isOpen: boolean;
}

export interface State {
  stations: Array<StationItem>
  currentlyPlaying: StationItem | null;
}

interface ActionTypes {
  setCurrentlyPlaying: string
  setStations: string
}

export type Action = {
  type: string;
  payload: {
    stations?: StationItem[];
    station?: StationItem;
  };
}

const initialState: State = {
  stations: [
    {name: 'Putin FM', id: '66,6', isOpen: false},
    {name: 'Dribbble FM', id: '101,2', isOpen: false},
    {name: 'Doge FM', id: '99,4', isOpen: false},
    {name: 'Ballads FM', id: '87,1', isOpen: false},
    {name: 'Maximum FM', id: '142,2', isOpen: false},
  ],
  currentlyPlaying: null,
}

const actionTypes: ActionTypes = {
  setStations: 'SET_STATIONS',
  setCurrentlyPlaying: 'SET_CURRENTLY_PLAYING',
}

export default function stationsReducer(state: State = initialState, action: Action) {
  switch (action.type) {
    case actionTypes.setStations:    
      return {
        ...state,
        stations: action?.payload?.stations,
      };
    case actionTypes.setCurrentlyPlaying:
      const copyStations = state.stations;

      const currentlyPlayingStation = copyStations.find(item => item.isOpen) as StationItem;
      if (currentlyPlayingStation 
        && currentlyPlayingStation?.id !== action.payload.station?.id
      ) {
        currentlyPlayingStation.isOpen = false;
        copyStations[copyStations.findIndex(el => el.id === currentlyPlayingStation?.id)] = currentlyPlayingStation;
      }

      const objIndex = copyStations.findIndex(((obj: StationItem) => obj.id === action.payload?.station?.id));
      copyStations[objIndex].isOpen = action?.payload?.station?.isOpen as boolean;
      return {
        stations: copyStations,
      };
    default:
      return state
  }
}
