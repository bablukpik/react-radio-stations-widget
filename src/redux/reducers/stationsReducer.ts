import { setCurrentlyPlaying } from '../../utils/helperFunctions';

export interface StationItems {
  name: string
  number: string
  isOpen: boolean
}

export interface State {
  stations: Array<StationItems>
  currentlyPlaying: string | null
}

interface ActionTypes {
  setCurrentlyPlaying: string
  getStations: string
  updateStations: string
}

export type Action = {
  type: string
  payload: string | boolean | Array<StationItems> | null
}

const initialState: State = {
  stations: [
    {name: 'Putin FM', number: '66,6', isOpen: false},
    {name: 'Dribbble FM', number: '101,2', isOpen: false},
    {name: 'Doge FM', number: '99,4', isOpen: false},
    {name: 'Ballads FM', number: '87,1', isOpen: false},
    {name: 'Maximum FM', number: '142,2', isOpen: false},
  ],
  currentlyPlaying: null,
}

const actionTypes: ActionTypes = {
  getStations: 'GET_STATIONS',
  updateStations: 'UPDATE_STATIONS',
  setCurrentlyPlaying: 'SET_CURRENTLY_PLAYING',
}

export default function stationsReducer(state = initialState, action: Action) {
  switch (action.type) {
    case actionTypes.getStations:
      return state;
    case actionTypes.setCurrentlyPlaying:
      if (state.currentlyPlaying === action.payload) {
        return state;
      }
      return {
        ...state,
        stations: setCurrentlyPlaying(state, action),
        currentlyPlaying: action.payload,
      };
    default:
      return state
  }
}
