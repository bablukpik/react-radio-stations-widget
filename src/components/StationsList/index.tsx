import React from 'react';
import { useDispatch } from 'react-redux';
import {useSpring, animated} from 'react-spring';
import styles from './StationsList.module.css';
import ListItem from '../ListItem';
import Text from '../Text';
import { State, StationItem } from '../../reducers/stationsReducer';
import Button from '../Button';

interface Props {
  state: State;
}

const StationList = ({ state }: Props) => {
  const { stations } = state;
  const {stationListWrapper, stationList} = styles
  const animateProps = useSpring({delay: 100, opacity: 1, from: {opacity: 0}})

  const dispatch = useDispatch()
  const handleClick = (item: StationItem) => {
    dispatch({
      type: 'SET_CURRENTLY_PLAYING',
      payload: { 
        station: { ...item, isOpen: !item.isOpen },
      },
    });
  }
  return (
    <div className={stationListWrapper}>
      <animated.ul className={stationList} style={animateProps}>
        {stations.map((item: StationItem) => {
          return (
            <React.Fragment key={item.id}>
              <ListItem
                key={item.id}
                isOpen={item.isOpen}
              >
                <Button onClick={() => handleClick(item)}>
                  <Text>{item.name}</Text>
                </Button>
                <Button onClick={() => handleClick(item)}>
                  <Text bold>{item.id}</Text>
                </Button>
              </ListItem>
            </React.Fragment>
          )
        })}
      </animated.ul>
    </div>
  )
}

export default StationList;
