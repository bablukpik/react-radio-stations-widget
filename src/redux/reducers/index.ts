import { combineReducers } from "redux";
import stationsReducer from './stationsReducer';
const reducers = combineReducers({
  allStations: stationsReducer,
});
export default reducers;
