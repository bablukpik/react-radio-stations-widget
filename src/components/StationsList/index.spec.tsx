import { FC } from 'react';
import { render } from '@testing-library/react';
import { initialState } from 'src/reducers/stationsReducer';

import StationsList from './index';
import store from 'src/store';
import { Provider } from 'react-redux';

const Wrapper: FC = ({ children }) => (
  <Provider store={store}>
    {children}
  </Provider>
);

describe('StationsList', () => {
  it('should render', () => {
    const { asFragment } = render(<StationsList state={initialState}/>, { wrapper: Wrapper });
    expect(asFragment()).toMatchSnapshot();
  });
});
