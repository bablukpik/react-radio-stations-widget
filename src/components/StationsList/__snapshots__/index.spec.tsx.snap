// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StationsList should render 1`] = `
<DocumentFragment>
  <div
    class="stationListWrapper"
  >
    <ul
      class="stationList"
      style="opacity: 0;"
    >
      <li
        class="listItem"
        id="66,6"
      >
        <span
          class="text"
        >
          Putin FM
        </span>
        <span
          class="text textBold"
        >
          66,6
        </span>
      </li>
      <li
        class="listItem"
        id="101,2"
      >
        <span
          class="text"
        >
          Dribbble FM
        </span>
        <span
          class="text textBold"
        >
          101,2
        </span>
      </li>
      <li
        class="listItem"
        id="99,4"
      >
        <span
          class="text"
        >
          Doge FM
        </span>
        <span
          class="text textBold"
        >
          99,4
        </span>
      </li>
      <li
        class="listItem"
        id="87,1"
      >
        <span
          class="text"
        >
          Ballads FM
        </span>
        <span
          class="text textBold"
        >
          87,1
        </span>
      </li>
      <li
        class="listItem"
        id="142,2"
      >
        <span
          class="text"
        >
          Maximum FM
        </span>
        <span
          class="text textBold"
        >
          142,2
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
